Kitodo-DataAccess/src/main/java/org/kitodo/dataaccess/format/xml/XMLWriter.java

```diff
@@ -0,0 +1,340 @@
+ */
+
+package org.kitodo.dataaccess.format.xml;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.TreeMap;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import org.kitodo.dataaccess.AccessibleObject;
+import org.kitodo.dataaccess.IdentifiableNode;
+import org.kitodo.dataaccess.LangString;
+import org.kitodo.dataaccess.Literal;
+import org.kitodo.dataaccess.Node;
+import org.kitodo.dataaccess.NodeReference;
+import org.kitodo.dataaccess.ObjectType;
+import org.kitodo.dataaccess.RDF;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+/**
+ * Transforms a linked data structure to an XML file.
+ *
+ * @author Matthias Ronge
+ */
+public class XMLWriter {
+    /**
+     * String to pass to the transformer to switch on indenting.
+     */
+    private static final String TRANSFORMER_INDENT_TRUE = "yes";
+
+    /**
+     * String to pass to the transformer to set the indenting width in
+     * characters.
+     */
+    private static final String TRANSFORMER_INDENT_VALUE = "{http://xml.apache.org/xslt}indent-amount";
+
+    /**
+     * Returns the type of the node, or {@code rdf:Description} if not set.
+     *
+     * @param node
+     *            node whose type shall be returned
+     * @return the type of the node, or {@code rdf:Description}
+     */
+    private static String getTypeOrDefault(AccessibleObject node) {
+        try {
+            return node.getType();
+        } catch (NoSuchElementException e) {
+            return RDF.DESCRIPTION;
+        }
+    }
+
+    /**
+     * Populates a DOM element with the data from this node.
+     *
+     * @param node
+     *            node to transform
+     * @param out
+     *            element to populate
+     * @param xml
+     *            the dom document, needed to create the children
+     * @param abbr
+     *            the namespace handler initialized at the root node
+     * @throws NoSuchElementException
+     *             if a relation links to an empty set of objects. (If so, the
+     *             object model is corrupt.)
+     */
+    private static void nodeToElement(Node node, Document xml, Namespaces abbr, Element out) {
```

Comment:
JavaDoc and method signature did not match. Please update the JavaDoc!

Reply:
Should be correct. `NoSuchElementException` is an unchecked exception that does not need to be declared. It is only explained in the Javadoc because the reason is not obvious if it happens. It may be thrown by `objects.iterator().next()` if the data model in memory is corrupted.

Reply:
On many methods you are throwing unchecked exceptions (most of them are RuntimeExceptions) without any explanation in JavaDoc and here you added it? I can not follow your minds in this kind of writing code and documentation.

Reply:
In this case, it has value (may help in the future) to understand what is going on, and the comment is already there. I don’t see a benefit in removing sensible documentation. In the Oracle JDK it is widespread practice to document runtime exceptions if they are important to a method, and otherwise not. I try to follow this practice. (Of course, there is *always* more that could be documented—feel free to do so.)

The `RuntimeException` cases are most likely cases where I wrap a checked exception that is the result of an either bad, or unconventionally used design. If you want to [delve deeper into this matter, I can really recommend you this article](http://www.onjava.com/pub/a/onjava/2003/11/19/exceptions.html#article).

**TL;DR**
* Checked exceptions should be used in cases where the *user* has to fix the problem and the front-end should generate a sensible error message. Common example: disk full.
* Unchecked exceptions should be used if the exception is the result of a programming error elsewhere, typically `NullPointerException` or `IllegalArgumentException`. Here, a *programmer* needs to fix it. A stacktrace in the front-end will be sufficient in this case, as the user cannot help it anyway.
* In the early days of Java some programmers did not fully understand the concepts behind checked and unchecked exceptions, not even always at Sun. If a library was implemented badly and throws checked exceptions where they are inappropriate, or in a case where a checked exception can be considered to be never thrown (like: `string.getBytes("UTF-8")` → throws checked UnsupportedEncodingException, but will never, since UTF-8 is built-in. The exception makes sense if the enconding string is *entered* manually by the user, but not if it is a constant that is known to work) best practice is to `catch (WhateverCheckedException e) { throw new RuntimeException(e.getMessage(), e); }` so the desing is clean, but *if* the exception ever happens, it is not lost. If it will ever happen, this is the class ‘programming error’ from above, because it was not previewn to be happening.